debug/scripts: introduce a set of scripts for collecting info

rouming opened the pull request:
Commit introduces a set of scripts, which helps to retrieve useful debug information.

iommu-groups.sh
--------------------------
The script outputs PCI devices grouped by the IOMMU group
    
qemu-affinities.sh
--------------------------
The script outputs affinities of QEMU threads
    
collect-info.sh
--------------------------
This script will be a part of 'debug' container and the main purpose of it is to collect all useful information about the EVE node running in one tar.gz file for further debugging. Ideally script is executed by the one, who maintains the edge node and experiences any problems which can be analysed by the EVE team. 

Script doesn't collect any secrets, but rather logs and system information.

The script should be executed from the debug container, so:

```
# eve enter debug (not needed if there is an SSH access)
# collect-info.sh
```

cc: @milan-zededa 
cc: @mikem-zed 
cc: @christoph-zededa 
cc: @shjala 

Folks, please take a look in the script, what else do we need to be collected?

OhmSpectator commented:
If we collect all the info that can be useful to debug any kind of issues, I would be happy to see the info about the CPUs affinities...
```
for pid in $(pgrep qemu) ; do echo "======= QEMU $pid threads: =======" ;  for spid in $(ps -T -o spid= -p "$pid" ) ;do taskset -pc "$spid" ; done ; done
```
The only problem is that the debug container does not include a correct version of the ps tool. I used to run `apk add procps` first. That installs the corresponding alpine package.

rouming commented:
> If we collect all the info that can be useful to debug any kind of issues, I would be happy to see the info about the CPUs affinities...
> 
> ```
> for pid in $(pgrep qemu) ; do echo "======= QEMU $pid threads: =======" ;  for spid in $(ps -T -o spid= -p "$pid" ) ;do taskset -pc "$spid" ; done ; done
> ```
> 
> The only problem is that the debug container does not include a correct version of the ps tool. I used to run `apk add procps` first. That installs the corresponding alpine package.

No issue to install it from the script. This is the debug container after all. Will include you snippet, thanks.

rouming commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -0,0 +1,44 @@
+lspci -vvv -t > $DIR/lspci-vvv-t
+lsblk -a      > $DIR/lsblk-a
+logread       > $DIR/logread
+
```
Yep. Thanks

rouming commented:
@rene forgot to ask you to take a look, so please

milan-zededa commented:
For networking state, EVE captures output of some Linux commands for each netdump: https://github.com/lf-edge/eve/blob/master/pkg/pillar/netdump/netdump.go#L218-L294
It might make sense to replicate it also inside this script.

rene commented:
> @rene forgot to ask you to take a look, so please

:+1:  I will take a look, and please, fix yetus errors...

rene commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -0,0 +1,44 @@
+cat /proc/slabinfo    > $DIR/slabinfo
+cat /proc/zoneinfo    > $DIR/zoneinfo
+cat /proc/mounts      > $DIR/mounts
+cat /proc/vmstat      > $DIR/vmstat
```
I think it's a good idea to also include:

To get more hardware information:
- /proc/cpuinfo
- output of dmidecode (not installed, can be added with apk, but it's also available on pillar container, so it should be present on hostfs file system)

Hypervisor config for VMs:
- /run/domainmgr/xen/*.cfg

Doesn't make sense to also include all json files from /run?
- I would do something like this: 
```shell
find /run -name "*.json" | tar -cvjf /tmp/json.tar.bz2 -T -
```

milan-zededa replied:
Roman has already included the whole /run directory, similarly to what we do for the USB stick "diag" method: https://github.com/lf-edge/eve/blob/master/pkg/pillar/scripts/device-steps.sh#L161-L174

rene replied:
That's true @milan-zededa, didn't see it, my bad. But still, I think we are only interested on json files, right? socket files will be ignored, touch files are empty and for pid we can grab from ps output... does it make sense copy the whole /run?

rouming replied:
@rene yeah, why not to copy the whole /run? 'socket ignored' is used to avoid tar warnings.  Will include the cpuinfo and dmicode.

rouming commented:
Difference to the previous version:
* Followed advicees from the folks, added more additional debug info
* Added iommu-groups.sh script for getting all pci devices grouped by the IOMMU
* Added qemu-affinities.sh script for getting affinities of QEMU threads (code taken from @OhmSpectator code snippet)

rouming commented:
Ok, Yetus, I'm ready!!!!

rouming commented:
Yetus, before you start, think about why you need all this. how much pain you bring to the people. think about it

rene commented on pkg/debug/scripts/qemu-affinities.sh:
```diff
@@ -0,0 +1,14 @@
+#
+
+# We need that for proper GNU version of the 'ps' tool
+apk add procps > /dev/null
```
I think script should fail here (like add "... || exit 1" or do it better with an error message) if procps cannot be installed (because it's essential for this script). Also, isn't safe to do a "apk update" before the apk add... ? I'm wondering when an EVE version get too old for the apk repos and someone try to run the script.... BTW, apk will depend on network connection, so it shouldn't be too difficult to get an error here...

rene replied:
You can also check if the package is installed with something like this:
```shell
apk list -I | grep ^procps
```

rene commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -0,0 +1,55 @@
+mkdir -p "$DIR"
+
+# Install usefull stuff
+apk add dmidecode
```
maybe add "apk update" before? See my other comment: https://github.com/lf-edge/eve/pull/3101/commits/98d4d950e8d65bb3f1a730f4e7ee84181f685194#r1137157900

rene commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -0,0 +1,55 @@
+chroot /hostfs lsusb -vvv -t > "$DIR/lsusb-vvv-t"
+
+dmesg         > "$DIR/dmesg"
+ps awux       > "$DIR/ps-awux"
```
If you are not installing ps here, it's not guaranteed that you have the GNU version instead of busybox, so it's better to remove these flags...

rene replied:
yes, but my point is that this ps doesn't consider any flag at all, so `ps flakeflags` will produce exactly the same output of `ps`. So it's better to keep only `ps > "$DIR/ps"` to avoid any confusion and/or feature issue (and keep it simple).

rouming replied:
True.

rouming commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -0,0 +1,44 @@
+chroot /hostfs lsusb -vvv -t > $DIR/lsusb-vvv-t
+
+dmesg         > $DIR/dmesg
+ps awux       > $DIR/ps-awux
```
They work, no worries.

rouming commented:
Difference to the previous version:
* Addressed comments
* Added networking part, @milan-zededa please take a look
